ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/KafkaConfigStore.java

```diff
@@ -0,0 +1,188 @@
+      this.ksqlProperties = ksqlProperties == null
+          ? Collections.emptyMap() : Collections.unmodifiableMap(ksqlProperties);
+    }
+
+    public Map<String, String> getKsqlProperties() {
+      return ksqlProperties;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      return o instanceof KsqlProperties
+          && Objects.equals(ksqlProperties, ((KsqlProperties) o).ksqlProperties);
+    }
+  }
+
+  private Optional<KsqlProperties> readProperties(
+      final String topicName,
+      final KafkaConsumer<String, KsqlProperties> consumer) {
+    final TopicPartition topicPartition = new TopicPartition(topicName, 0);
+    final List<TopicPartition> topicPartitionAsList = Collections.singletonList(topicPartition);
+
+    consumer.assign(topicPartitionAsList);
+    consumer.seekToBeginning(topicPartitionAsList);
+
+    final long endOffset = consumer.endOffsets(topicPartitionAsList).get(topicPartition);
+    while (consumer.position(topicPartition) < endOffset) {
+      log.debug(
+          "Reading from config topic. Position(%d) End(%d)",
+          consumer.position(topicPartition),
+          endOffset);
+      final ConsumerRecords<String, KsqlProperties> records
+          = consumer.poll(Duration.of(5, ChronoUnit.SECONDS));
+      final Optional<ConsumerRecord<String, KsqlProperties>> record =
+          records.records(topicPartition)
+              .stream()
+              .filter(r -> r.key().equals(CONFIG_MSG_KEY))
+              .findFirst();
+      if (record.isPresent()) {
+        log.info("Found existing config in config topic");
+        return Optional.of(record.get().value());
+      }
+    }
+    log.info("No config found on config topic");
+    return Optional.empty();
+  }
+
+  private void writeProperties(final String topicName,
+                               final KsqlProperties properties,
+                               final KafkaProducer<String, KsqlProperties> producer) {
+    producer.send(new ProducerRecord<>(topicName, CONFIG_MSG_KEY, properties));
+    producer.flush();
+  }
+
+  private Map<String, String> readMaybeWriteProperties(
+      final KsqlConfig ksqlConfig,
+      final KafkaTopicClient topicClient,
+      final KafkaConsumer<String, KsqlProperties> consumer,
+      final KafkaProducer<String, KsqlProperties> producer) {
+    final String topicName = KsqlInternalTopicUtils.getTopicName(ksqlConfig, CONFIG_TOPIC_SUFFIX);
+    log.info("Ensure existence of topic %s", topicName);
+    KsqlInternalTopicUtils.ensureTopic(
+        topicName, ksqlConfig, topicClient, TopicConfig.CLEANUP_POLICY_COMPACT);
+
+    final Optional<KsqlProperties> properties = readProperties(topicName, consumer);
+    if (properties.isPresent()) {
+      return properties.get().getKsqlProperties();
+    }
+
+    log.info("Writing current config to config topic");
+    writeProperties(topicName, new KsqlProperties(ksqlConfig.getAllConfigPropsCleaned()), producer);
```

Comment:
is this going to be a race condition between multiple instances? is it something that we care about? or are we assuming that they are all writing the same properties, so last one wins?